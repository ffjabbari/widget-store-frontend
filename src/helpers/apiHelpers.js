/** API Helper functions. */

// local imports
import { asyncAPIRequest } from '../hooks/apiHook';

/**
 * updateUserProfile()
 *
 * Make patch request to update a user profile. Used by CustomerAddressForm
 * to allow updating customer address of signed in users.
 *
 * @param {object} values - form values
 * @param {object} user - user object
 * @param {token} token - user token
 * @param {function} setResponseError - state setter for error display
 * @param {any} defaultAddress - Falsey or obj with address components
 * @param {function} setFormDisabled - state setter to hide form after update
 *
 * returns: true if an error happened else undefined
 */

export const updateUserProfile = async (
  values,
  user,
  token,
  setResponseError,
  defaultAddress,
  setFormDisabled
) => {
  const userUpdateData = { ...values };
  // remove email to dis-allow email updating from here
  delete userUpdateData['email'];
  delete userUpdateData['user_id'];

  // remomve key/value pairs having a value of empty string from userUpdateData
  Object.entries(userUpdateData).forEach(([key, value]) => {
    if (value === '') delete userUpdateData[key];
  });

  try {
    // make patch request
    const resp = await asyncAPIRequest(
      `/users/${user.username}`,
      'patch',
      userUpdateData,
      { _token: token }
    );
    if (resp.error) {
      setResponseError(resp.error.response.data.message);
      return true;
    } else {
      defaultAddress.current = values;
      setFormDisabled(true);
    }
  } catch (error) {
    setResponseError(error.message);
    return true;
  }
};
