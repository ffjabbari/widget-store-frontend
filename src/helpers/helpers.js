import { Redirect } from 'react-router-dom';
import { motion } from 'framer-motion';

/**
 * Given a path return the "root" argument.
 * For example the "root" of /shop/product/206 would be "/shop".
 *
 * @param {string} path
 */

export const getPathRoot = path => {
  return path.match(/^(\/\w*)\/?.*$/)[1];
};

export const FramerRedirect = props => (
  <motion.div exit="undefined">
    <Redirect {...props} />
  </motion.div>
);

export const animateVariant = (setterFunct, delay, callback, args = []) => {
  // if animation state is false set to 'default' else 'active'
  setterFunct(state => (state === false ? 'default' : 'active'));
  setTimeout(() => {
    setterFunct('default');
    if (callback) callback(...args);
  }, delay);
};
