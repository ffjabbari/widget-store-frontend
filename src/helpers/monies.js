import Decimal from 'decimal.js';

const SF_SALES_TAX = new Decimal('0.085');

// calculate discounted price
export function calculateDiscountPrice(product) {
  if (product.discount === '0.00') return product.price;
  const discount = new Decimal(product.discount),
    price = new Decimal(product.price);
  return price.minus(price.times(discount)).toFixed(2);
}

/** Calculate discount as a percent. */
export function calculateDiscountPercent(product) {
  const discount = new Decimal(product.discount);
  return discount.times(100).toFixed(0);
}

/** Return discounted price and discount as a percent */
export function getDiscountData(product) {
  return {
    discountPrice: calculateDiscountPrice(product),
    percent: calculateDiscountPercent(product),
  };
}

/** Calculate tax if customer is in California */
export function calculateTax(subtotal, state) {
  if (!/^ca(lifornia)?$/i.test(state)) return '0.00';
  subtotal = new Decimal(subtotal);
  return subtotal.times(SF_SALES_TAX).toFixed(2);
}

/** Calculate tax plus subtotal plus shipping cost if provided */
export function calculateTotal(subtotal, state, shippingCost) {
  const tax = calculateTax(subtotal, state);
  subtotal = new Decimal(subtotal);
  shippingCost = shippingCost || 0;
  return subtotal.plus(tax).plus(shippingCost).toFixed(2);
}
