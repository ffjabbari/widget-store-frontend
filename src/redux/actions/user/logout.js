/** Log user out action creator. */

import { LOG_OUT } from '../types';

function logout() {
  return { type: LOG_OUT };
}

export default logout;
