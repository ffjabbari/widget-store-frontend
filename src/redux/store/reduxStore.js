/*
 * Uses redux-persist to persist state.
 * Uses redux-thunk to allow async operations.
 */

import { composeWithDevTools } from 'redux-devtools-extension';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';
import autoMergeLevel2 from 'redux-persist/lib/stateReconciler/autoMergeLevel2';
import thunk from 'redux-thunk';
import { createStore, applyMiddleware } from 'redux';

// local imports
import root from '../reducers/root';

const persistConfig = {
  key: 'root',
  storage,
  // allow any new state properties to persist state rehydration
  stateReconciler: autoMergeLevel2,
};

const persistedReducer = persistReducer(persistConfig, root);

export const store = createStore(
  persistedReducer,
  composeWithDevTools(applyMiddleware(thunk))
);

export const persistedStore = persistStore(store);

export const testStore = createStore(root);
