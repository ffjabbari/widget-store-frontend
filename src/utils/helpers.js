import { Redirect, Link } from 'react-router-dom';
import { motion } from 'framer-motion';

/**
 * getPathRoot()
 *
 * Given a path return the "root" argument.
 * For example the "root" of /shop/product/206 would be "/shop".
 *
 * @param {string} path
 */

export const getPathRoot = path => {
  return path.match(/^(\/\w*)\/?.*$/)[1];
};

/**
 * <FramerRedirect to="" />
 *
 * Component to force react-router-dom redirect when framer-motion
 * prevents a redirect from executing.
 *
 * If regular <Redirect /> fails in this environment use this redirect instead.
 *
 * @param {any} props
 *
 * returns: motion.div wrapped redirect component
 */
export const FramerRedirect = props => (
  <motion.div exit="undefined">
    <Redirect {...props} />
  </motion.div>
);
export const FramerLink = props => (
  <motion.div exit="undefined">
    <Link {...props} />
  </motion.div>
);

/**
 * animateVariant()
 *
 * Animation function to turn on an animation state then turn it back
 * off after a delay. The state will be set "active" then back to "default".
 *
 * If inital state provided by setterFunct is false this function will skip
 * setting the animation state "active" instead set it to "default".
 *
 * If a callback is provided it will be called with any arguments
 * after setting state back to "default".
 *
 *
 * @param {function} setterFunct - useState setter funct to set animation state
 * @param {integer} delay - delay of setterFunct being called
 * @param {funct} callback - function to call upon completion
 * @param {array} args - array of arguments for callback function
 */

export const animateVariant = (setterFunct, delay, callback, args = []) => {
  // if animation state is false set to 'default' else 'active'
  setterFunct(state => (state === false ? 'default' : 'active'));
  // after delay set back to "default"
  return setTimeout(() => {
    setterFunct('default');
    if (callback) callback(...args);
  }, delay);
};
