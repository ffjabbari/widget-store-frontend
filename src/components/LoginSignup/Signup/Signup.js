import { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { Redirect, useLocation } from 'react-router-dom';
import { Formik, Form, Field, ErrorMessage } from 'formik';
import { Button, FormGroup, Label, Alert, FormText } from 'reactstrap';

// local imports
import { asyncAPIRequest } from '../../../hooks/apiHook';
import { getPathRoot } from '../../../helpers/helpers';
import login from '../../../redux/actions/user/login';

const Signup = ({ handleClose }) => {
  const user = useSelector(state => state.user.user);
  const [responseError, setResponseError] = useState(null);
  const dispatch = useDispatch();
  const location = useLocation();

  if (user) return <Redirect to={`${getPathRoot(location.pathname)}`} />;

  const handleSubmit = async values => {
    try {
      const resp = await asyncAPIRequest('/users', 'post', values);
      if (resp.error) {
        setResponseError(resp.error.response.data.message);
      } else {
        const { user, token } = resp.data;
        dispatch(login(user, token));
        return <Redirect to={`${getPathRoot(location.pathname)}`} />;
      }
    } catch (error) {
      console.log(error);
    }
  };

  return (
    <div className="SignupForm">
      <h1>Signup</h1>
      <Formik
        initialValues={{ username: '', email: '', password: '' }}
        validate={signupValidation}
        onSubmit={(values, { setSubmitting }) => {
          handleSubmit(values);
          setSubmitting(false);
        }}
      >
        {({ isSubmitting }) => (
          <Form className="text-left">
            <FormGroup>
              <Label htmlFor="username">Username</Label>
              <Field
                type="text"
                name="username"
                id="username"
                autoComplete="username"
                className="form-control"
              />
              <ErrorMessage
                name="username"
                component="div"
                className="text-danger mt-1"
              />
            </FormGroup>
            <FormGroup>
              <Label htmlFor="email2">Email</Label>
              <Field
                type="email"
                name="email"
                id="email2"
                autoComplete="email"
                className="form-control"
              />
              <ErrorMessage
                name="email"
                component="div"
                className="text-danger mt-1"
              />
            </FormGroup>
            <FormGroup>
              <Label htmlFor="password2">Password</Label>
              <Field
                type="password"
                name="password"
                id="password2"
                autoComplete="current-password"
                className="form-control"
              />
              <FormText>
                <ul className="pl-3">
                  <li>Minimum length 8 characters</li>
                  <li>
                    Required 1 uppercase letter, 1 lowercase letter, and 1
                    number
                  </li>
                </ul>
              </FormText>
              <ErrorMessage
                name="password"
                component="div"
                className="text-danger mt-1"
              />
            </FormGroup>
            {responseError && <Alert color="danger">{responseError}</Alert>}
            <Button type="submit" color="primary" disabled={isSubmitting}>
              Signup
            </Button>
            <Button color="secondary ml-2" onClick={handleClose}>
              Cancel
            </Button>
          </Form>
        )}
      </Formik>
    </div>
  );
};

/** Validation logic */
const signupValidation = values => {
  const errors = {};
  if (!values.username) {
    errors.username = 'Username Required';
  } else if (!/^[^   ]{2,55}$/.test(values.username)) {
    errors.username = 'Username must be between 2 and 55 characters';
  }
  if (!values.email) {
    errors.email = 'Email Required';
  } else if (!/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,}$/i.test(values.email)) {
    errors.email = 'Invalid email address';
  }
  if (!values.password) {
    errors.password = 'Password Required';
  } else if (
    !/^(?=.*\d)(?=.*[a-z])(?=.*[A-Z])(?=.*[a-zA-Z]).{8,55}$/.test(
      values.password
    )
  ) {
    errors.password =
      'Password must be at least 8 characters long, contain at least 1 uppercase letter, 1 lowercase letter, and 1 number';
  }
  return errors;
};

export default Signup;
