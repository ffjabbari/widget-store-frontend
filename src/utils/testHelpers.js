import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { testStore } from '../redux/store/reduxStore';

export const renderWithStore = component =>
  render(
    <Provider store={testStore}>
      <MemoryRouter>{component}</MemoryRouter>
    </Provider>
  );
