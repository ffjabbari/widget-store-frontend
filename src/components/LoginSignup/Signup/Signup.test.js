import { act, fireEvent } from '@testing-library/react';
import axios from 'axios';

// local imports
import { renderWithStore } from '../../../helpers/testHelpers';
import { testStore } from '../../../redux/store/reduxStore';
import logout from '../../../redux/actions/user/logout';
import Signup from './Signup';

jest.mock('axios');

afterEach(() => {
  jest.resetAllMocks();
});

test('should render', () => {
  renderWithStore(<Signup />);
});

test('Signup snapshot', async () => {
  const { asFragment } = renderWithStore(<Signup />);
  expect(asFragment()).toMatchSnapshot();
});

test('shoud store user data in store with submit success', async () => {
  axios.mockResolvedValue({ data: { user: { user: 'user' }, token: 'token' } });

  const { getByLabelText, getByRole } = renderWithStore(<Signup />);

  const usernameInput = getByLabelText(/username/i);
  await act(async () => {
    fireEvent.change(usernameInput, { target: { value: 'test_user' } });
  });

  const emailInput = getByLabelText(/email/i);
  await act(async () => {
    fireEvent.change(emailInput, { target: { value: 'abc@def.gh' } });
  });

  const passwordInput = getByLabelText(/password/i);
  await act(async () => {
    fireEvent.change(passwordInput, { target: { value: 'Password1' } });
  });

  const SignupButton = getByRole('button', { name: /signup/i });
  await act(async () => {
    fireEvent.click(SignupButton);
  });

  expect(axios).toHaveBeenCalled();
  expect(axios.mock.calls.length).toBe(1);
  expect(axios.mock.calls[0][0].url).toBe('/users');
  expect(axios.mock.calls[0][0].method).toBe('post');
  expect(axios.mock.calls[0][0].data.username).toBe('test_user');
  expect(axios.mock.calls[0][0].data.email).toBe('abc@def.gh');
  expect(axios.mock.calls[0][0].data.password).toBe('Password1');

  // expect store.user data to have been populated
  const user = testStore.getState().user;
  expect(user.token).toBe('token');
  expect(user.user.user).toBe('user');

  testStore.dispatch(logout());
});

test('shoud not call axios with bad username', async () => {
  const { getByLabelText, getByRole } = renderWithStore(<Signup />);

  const usernameInput = getByLabelText(/username/i);
  await act(async () => {
    fireEvent.change(usernameInput, { target: { value: 't' } });
  });

  const emailInput = getByLabelText(/email/i);
  await act(async () => {
    fireEvent.change(emailInput, { target: { value: 'abc@def.gh' } });
  });

  const passwordInput = getByLabelText(/password/i);
  await act(async () => {
    fireEvent.change(passwordInput, { target: { value: 'Password1' } });
  });

  const SignupButton = getByRole('button', { name: /Signup/i });
  await act(async () => {
    fireEvent.click(SignupButton);
  });

  expect(axios).not.toHaveBeenCalled();
});

test('shoud not call axios with bad email', async () => {
  const { getByLabelText, getByRole } = renderWithStore(<Signup />);

  const usernameInput = getByLabelText(/username/i);
  await act(async () => {
    fireEvent.change(usernameInput, { target: { value: 'test_user' } });
  });

  const emailInput = getByLabelText(/email/i);
  await act(async () => {
    fireEvent.change(emailInput, { target: { value: 'abc@def.' } });
  });

  const passwordInput = getByLabelText(/password/i);
  await act(async () => {
    fireEvent.change(passwordInput, { target: { value: 'Password1' } });
  });

  const SignupButton = getByRole('button', { name: /Signup/i });
  await act(async () => {
    fireEvent.click(SignupButton);
  });

  expect(axios).not.toHaveBeenCalled();
});

test('shoud not call axios with bad password', async () => {
  const { getByLabelText, getByRole } = renderWithStore(<Signup />);

  const usernameInput = getByLabelText(/username/i);
  await act(async () => {
    fireEvent.change(usernameInput, { target: { value: 'test_user' } });
  });

  const emailInput = getByLabelText(/email/i);
  await act(async () => {
    fireEvent.change(emailInput, { target: { value: 'abc@def.gh' } });
  });

  const passwordInput = getByLabelText(/password/i);
  await act(async () => {
    fireEvent.change(passwordInput, { target: { value: 'Password' } });
  });

  const SignupButton = getByRole('button', { name: /Signup/i });
  await act(async () => {
    fireEvent.click(SignupButton);
  });

  expect(axios).not.toHaveBeenCalled();
});

test('shoud show error message with submit failure', async () => {
  axios.mockResolvedValue({
    error: { response: { data: { message: 'error message' } } },
  });

  const { getByLabelText, getByRole } = renderWithStore(<Signup />);

  const usernameInput = getByLabelText(/username/i);
  await act(async () => {
    fireEvent.change(usernameInput, { target: { value: 'test_user' } });
  });

  const emailInput = getByLabelText(/email/i);
  await act(async () => {
    fireEvent.change(emailInput, { target: { value: 'abc@def.gh' } });
  });

  const passwordInput = getByLabelText(/password/i);
  await act(async () => {
    fireEvent.change(passwordInput, { target: { value: 'Password1' } });
  });

  const SignupButton = getByRole('button', { name: /Signup/i });
  await act(async () => {
    fireEvent.click(SignupButton);
  });

  expect(axios).toHaveBeenCalled();
  const user = testStore.getState().user;

  expect(user.token).toBe(null);
  expect(user.user).toBe(null);

  const errorMessage = getByRole('alert');
  expect(errorMessage).toBeInTheDocument();
  expect(errorMessage.textContent).toBe('error message');
});
