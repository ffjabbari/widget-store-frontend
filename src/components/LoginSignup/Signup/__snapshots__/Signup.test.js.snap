// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup snapshot 1`] = `
<DocumentFragment>
  <div
    class="SignupForm"
  >
    <h1>
      Signup
    </h1>
    <form
      action="#"
      class="text-left"
    >
      <div
        class="form-group"
      >
        <label
          class=""
          for="username"
        >
          Username
        </label>
        <input
          autocomplete="username"
          class="form-control"
          id="username"
          name="username"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class=""
          for="email2"
        >
          Email
        </label>
        <input
          autocomplete="email"
          class="form-control"
          id="email2"
          name="email"
          type="email"
          value=""
        />
      </div>
      <div
        class="form-group"
      >
        <label
          class=""
          for="password2"
        >
          Password
        </label>
        <input
          autocomplete="current-password"
          class="form-control"
          id="password2"
          name="password"
          type="password"
          value=""
        />
        <small
          class="form-text text-muted"
        >
          <ul
            class="pl-3"
          >
            <li>
              Minimum length 8 characters
            </li>
            <li>
              Required 1 uppercase letter, 1 lowercase letter, and 1 number
            </li>
          </ul>
        </small>
      </div>
      <button
        class="btn btn-primary"
        type="submit"
      >
        Signup
      </button>
      <button
        class="btn btn-secondary ml-2"
      >
        Cancel
      </button>
    </form>
  </div>
</DocumentFragment>
`;
